src/coreclr/src/vm/dllimport.cpp

```diff
@@ -6532,476 +6473,603 @@ INT_PTR NDirect::GetNativeLibraryExport(NATIVE_LIBRARY_HANDLE handle, LPCWSTR sy
+            {
+                hmod = LoadFromPInvokeAssemblyDirectory(callingAssembly, currLibNameVariation, loadWithAlteredPathFlags | dllImportSearchPathFlags, pErrorTracker);
+                if (hmod != NULL)
+                {
+                    return hmod;
+                }
             }
 
-            hmod = LocalLoadLibraryHelper(currLibNameVariation, flags, pErrorTracker);
+            hmod = LocalLoadLibraryHelper(currLibNameVariation, dllImportSearchPathFlags, pErrorTracker);
             if (hmod != NULL)
             {
                 return hmod;
             }
         }
-        else if ((callingAssembly != nullptr) && searchAssemblyDirectory)
+
+        // This may be an assembly name
+        // Format is "fileName, assemblyDisplayName"
+        MAKE_UTF8PTR_FROMWIDE(szLibName, wszLibName);
+        char *szComma = strchr(szLibName, ',');
+        if (szComma)
         {
-            hmod = LoadFromPInvokeAssemblyDirectory(callingAssembly, currLibNameVariation, loadWithAlteredPathFlags | dllImportSearchPathFlags, pErrorTracker);
-            if (hmod != NULL)
+            *szComma = '\0';
+            // Trim white spaces
+            while (COMCharacter::nativeIsWhiteSpace(*(++szComma)));
+
+            AssemblySpec spec;
+            if (SUCCEEDED(spec.Init(szComma)))
             {
-                return hmod;
+                // Need to perform case insensitive hashing.
+                SString moduleName(SString::Utf8, szLibName);
+                moduleName.LowerCase();
+
+                StackScratchBuffer buffer;
+                szLibName = (LPSTR)moduleName.GetUTF8(buffer);
+
+                Assembly *pAssembly = spec.LoadAssembly(FILE_LOADED);
+                Module *pModule = pAssembly->FindModuleByName(szLibName);
+
+                hmod = LocalLoadLibraryHelper(pModule->GetPath(), loadWithAlteredPathFlags | dllImportSearchPathFlags, pErrorTracker);
             }
         }
 
-        hmod = LocalLoadLibraryHelper(currLibNameVariation, dllImportSearchPathFlags, pErrorTracker);
-        if (hmod != NULL)
+        return hmod;
+    }
+
+    NATIVE_LIBRARY_HANDLE LoadLibraryModuleBySearch(NDirectMethodDesc *pMD, LoadLibErrorTracker *pErrorTracker, PCWSTR wszLibName)
+    {
+        STANDARD_VM_CONTRACT;
+
+        BOOL searchAssemblyDirectory;
+        DWORD dllImportSearchPathFlags;
+
+        GetDllImportSearchPathFlags(pMD, &dllImportSearchPathFlags, &searchAssemblyDirectory);
+
+        Assembly *pAssembly = pMD->GetMethodTable()->GetAssembly();
+        return LoadLibraryModuleBySearch(pAssembly, searchAssemblyDirectory, dllImportSearchPathFlags, pErrorTracker, wszLibName);
+    }
+
+    // This thread local and UseExtensionPoints class are used to guard against infinite recursion.
+    // The managed APIs (NativeLibarary.Load/TryLoad) that invoke managed extension points can also
+    // be called from within the extension points themselves. This class tracks whether or not the
```

Comment:
If the call has different argument, then we should allow. There are valid cases where somebody may want to do that.

Reply:
BTW: We also have an option to "do nothing" and see whether we hear any reports of breaks due to infinite recursion.

Reply:
I think the 'do nothing' option becomes much more appealing if we avoid calling into the `DllImportResolver` (since that seems like the more likely to have breaks based on usage/searching around). Especially with differentiating the arguments, it seems like the extra tracking may not be worth doing without knowing it will be an issue (and easy enough to add if there is).

Reply:
Removed the tracking/blocking and only going through the extension points tied to the ALC now.